crates/emitter/src/control_structures.rs

```diff
@@ -586,13 +657,67 @@ where
     F1: Fn(&mut AstEmitter) -> Result<(), EmitError>,
 {
     pub fn emit(&mut self, emitter: &mut AstEmitter) -> Result<(), EmitError> {
-        emitter
-            .control_stack
-            .open_label(self.name, &mut emitter.emit);
+        emitter.control_stack.open_label(
+            self.name,
+            &mut emitter.emit,
+            self.enclosing_emitter_scope_index,
+        );
 
         (self.body)(emitter)?;
 
         emitter.control_stack.close_label(&mut emitter.emit);
         Ok(())
     }
 }
+
+pub enum NonLocalExitKind {
+    Continue,
+    Break,
+}
+
+pub struct NonLocalExitControl {
+    pub kind: NonLocalExitKind,
+}
+
+impl NonLocalExitControl {
+    pub fn prepare_for_non_local_jump(
+        &mut self,
+        emitter: &mut AstEmitter,
+        label: Option<SourceAtomSetIndex>,
+    ) -> Vec<ScopeNoteHoleEmitter> {
+        let enclosing_emitter_scope_index = match label {
+            Some(label) => emitter
+                .control_stack
+                .find_labelled_control(label)
+                .enclosing_emitter_scope_index(),
+            None => emitter
+                .control_stack
+                .innermost()
+                .enclosing_emitter_scope_index(),
+        };
+
+        let current_scope_index = emitter.scope_stack.current_index();
+
+        let mut holes = Vec::new();
+        let scope_indicies = emitter
+            .scope_stack
+            .scope_note_indices_from_to(&enclosing_emitter_scope_index, &current_scope_index);
+        for maybe_scope_note_index in scope_indicies.iter().rev() {
+            holes.push(ScopeNoteHoleEmitter {
+                scope_note_index: emitter.emit.enter_scope_hole(maybe_scope_note_index),
+            });
+        }
+
+        holes
+    }
+}
+
+pub struct ScopeNoteHoleEmitter {
+    scope_note_index: ScopeNoteIndex,
+}
+
+impl ScopeNoteHoleEmitter {
+    pub fn emit_non_local_jump_end(&self, emit: &mut InstructionWriter) {
+        emit.leave_scope_hole(self.scope_note_index);
+    }
+}
```

Comment:
instead of returning a vec of this struct and let the caller call this method for each,
it would be nice to create a struct that holds the vec of indices, and let the caller call single method of it, that calls `leave_scope_hole` for each index.

Or, making it more similar to other emitter helpers,
`NonLocalExitControl` can take a closure that is performed inside scope hole,
and enter and leave inside single method, so that the vec of indices aren't exposed to the caller,
and there's no possibility to forget to leave the hole.